### PR TITLE
Sync simplecpp::simplifyPath() with cppcheck's Path::simplifyPath()

### PR DESCRIPTION
**Pull request by amai2012**
As suggested this PR aims to synchronize the simplifyPath() routine from cppcheck and simplecpp.

So far tests from cppcheck have been imported to simplecpp, a few have been added on top. Currently on Linux 4 are still failing, number on Windows is unknown and will be addressed later.

### CODE REVIEW

File: simplecpp.cpp [hunk 1]
```diff
@@ -1724,11 +1724,31 @@ namespace simplecpp {
      */
     std::string simplifyPath(std::string path)
     {
+        const bool isUnc = path.compare(0,2,"//") == 0;
+
+        // Remove ./, .//, ./// etc. at the beginning
+        while (path.compare(0,2,"./") == 0) { // remove "./././"
```
Comment (danmar) [inline simplecpp.cpp]: this should be after `// replace backslash separators` so .\ is handled the same way

File: simplecpp.cpp [hunk 2]
```diff
@@ -1724,11 +1724,31 @@ namespace simplecpp {
      */
     std::string simplifyPath(std::string path)
     {
+        const bool isUnc = path.compare(0,2,"//") == 0;
```
Comment (danmar) [inline simplecpp.cpp]: is `\\` unc?

**Comment by danmar** [comment]
I assume this is a work-in-progress PR. It looks promising!

File: simplecpp.cpp (hunk 2 above)
Comment (danmar) [inline simplecpp.cpp]: sorry.. unc is always "\\...." right? then this is the best location.

**Comment by amai2012** [comment]
Yes indeed, this is work-in-progress, we shouldn't apply this yet.
And we should make sure that the expected/wanted test results are really ok, before changing the implementation...

File: simplecpp.cpp [hunk 3]
```diff
@@ -1716,66 +1716,123 @@ namespace simplecpp {
         return ostr.str();
     }
 #else
-#define realFilename(f)  f
+    // Boring implementation, but could help in platform-independent tests:
```
Comment (danmar) [inline simplecpp.cpp]: A better comment would be something like: `for case sensitive file systems`

File: simplecpp.cpp [hunk 4]
```diff
@@ -1716,66 +1716,123 @@ namespace simplecpp {
...
+    std::string realFilename(const std::string &f)
```
Comment (danmar) [inline simplecpp.cpp]: this function do stuff, copy the string.

File: simplecpp.cpp (hunk 3 above)
Comment (amai2012) [inline simplecpp.cpp]: Well, the Windows implementation does more than this...

**Comment by amai2012** [comment]
Maybe, the code is longer but better to understand imho (ok, I wrote it...)

What about the test results? Are the changes reasonable? Tests for Windows/DOS paths including drive letters are still missing.

File: simplecpp.cpp (hunk 3 above)
Comment (danmar) [inline simplecpp.cpp]: As far as I know realFilename just changes the case so it matches the filesystem. Do you see any additional behaviour?

File: simplecpp.cpp (hunk 4 above)
Comment (danmar) [inline simplecpp.cpp]: The macro will not do anything so it is cheaper

**Comment by danmar** [comment]
I think it might be a good idea to handle currentPath also.

If currentPath is  /123/456 then paths should be simplified like:

../456 => . 
../abc  => ../abc

To allow unit tests it can for instance be a parameter.

File: simplecpp.cpp (hunk 4 above)
Comment (amai2012) [inline simplecpp.cpp]: Yes, but as I wrote one might write tests only if there is a real function, cannot write tests for the macro.

Comment (danmar) [inline simplecpp.cpp]: it would be nice to test realFilename in windows. since there has been problems. but I am skeptic about testing it with unit tests since it needs files on the disk.

I don't say you should implement that now... but maybe a separate test script for windows could test it.

File: simplecpp.cpp [hunk 5]
```diff
@@ -1716,42 +1716,127 @@ namespace simplecpp {
...
 
-        // remove "./"
-        pos = 0;
-        while ((pos = path.find("./",pos)) != std::string::npos) {
-            if (pos == 0 || path[pos - 1U] == '/')
-                path.erase(pos,2);
-            else
-                pos += 2;
+    /// split path string into components - neglecting the path separators!
+    static PathPartsT SplitPath(const std::string& path)
```
Comment (danmar) [inline simplecpp.cpp]: nit: first function name letter should be lowercase.

File: simplecpp.cpp [hunk 6]
```diff
@@ -1716,42 +1716,127 @@ namespace simplecpp {
...
-                path.erase(0,pos+4);
-                pos = 0;
-            } else {
-                path.erase(pos1,pos-pos1+3);
-                pos = std::min((std::string::size_type)1, pos1);
+        return pathParts;
+    }
+
+    // Normalize path: resolve .., remove meaningless .
+    static void NormalizePath(PathPartsT& pathParts)
```
Comment (danmar) [inline simplecpp.cpp]: nit: first function name letter should be lowercase.

File: simplecpp.cpp [hunk 7]
```diff
@@ -1716,42 +1716,127 @@ namespace simplecpp {
...
+    std::string simplifyPath(std::string path)
+    {
+        if (path.empty())
+            return "";
+        // detect UNC path on DOS/Windows
+        const bool isUnc = (0==path.compare(0,2,"\\\\")) || (0==path.compare(0,2,"//"));
+        // replace backslash separators
+        std::replace(path.begin(), path.end(), '\\', SEP);
+        // detect: DOS/Windows path including drive letter
+        const bool isAbsolute=path[0]==SEP;
```
Comment (danmar) [inline simplecpp.cpp]: I think `isAbsolute` could be `(path[0]==SEP || DOSDriveLetterPathAbsolute(path))` without changing the meaning.

File: simplecpp.cpp [hunk 8]
```diff
@@ -1716,42 +1716,127 @@ namespace simplecpp {
...
+        return f;
+    }
 #endif
 
-    /**
-     * perform path simplifications for . and ..
-     */
-    std::string simplifyPath(std::string path)
+    static const char SEP='/';
+    static bool DOSDriveLetterPath(const std::string& path)
```
Comment (danmar) [inline simplecpp.cpp]: as far as I see this is only used in the `DOSDriveLetterPathAbsolute`.. I would integrate this.

File: test.cpp [hunk 9]
```diff
@@ -1,28 +1,57 @@
...
         std::cerr << "expected:" << expected << std::endl;
         std::cerr << "actual:" << actual << std::endl;
     }
     return (expected == actual);
 }
 
-static int assertEquals(const unsigned int &expected, const unsigned int &actual, int line)
+static int succeeded_todos_counter;
+static int todos_counter;
+static int todoAssertEquals(const std::string &wanted, const std::string &expected, const std::string &actual, const std::string & function, const std::string &file, int line)
```
Comment (danmar) [inline test.cpp]: I would like that this is implemented in a separate PR. just add the test cases. however I don't currently see a need for TODO_ASSERT.

File: test.cpp [hunk 10]
```diff
@@ -1257,6 +1286,65 @@ void simplifyPath()
     ASSERT_EQUALS("/../1.c", simplecpp::simplifyPath("/../a/../1.c"));
 }
 
+// tests transferred from cppcheck
+// https://github.com/danmar/cppcheck/blob/d3e79b71b5ec6e641ca3e516cfced623b27988af/test/testpath.cpp#L43 ff.
+void simplifyPath_cppcheck()
```
Comment (danmar) [inline test.cpp]: as far as I see all these test cases look good

Comment (danmar) [inline test.cpp]: I did not see the "sorting" of the testcases making it hard to see what corner cases are covered. or if there are duplicate tests.

do we need all tests with // ? that is 1 condition in the code that is tested with 17 assertions. if those are needed I would suggest we group all equivalent tests together. a/b a//b a///b are put together. then when reading those one does not need to re-interpret the path every time.

**Comment by danmar** [comment]
I implemented most of this functionality with eff31dc75db2c7295177158e59e4e91f647a45b2. however as you say, tests with dos drive letters are still missing. I plan to look at that soon.
